rust/src/sans_io/indexed_reader.rs

```diff
@@ -0,0 +1,872 @@
+        }
+        // we're out of message indexes, we need to load a chunk.
+        // if we're out of chunks, we're done.
+        if self.cur_chunk_index >= self.chunk_indexes.len() {
+            return None;
+        }
+        chunk_request(&self.chunk_indexes[self.cur_chunk_index])
+    }
+
+    /// Call to insert new compressed records into this reader. `offset` should be a valid file
+    /// offset to the start of the compressed data in a chunk. `compressed_data` should be a slice
+    /// containing the entire compressed data for that chunk.
+    pub fn insert_chunk_record_data(
+        &mut self,
+        offset: u64,
+        compressed_data: &[u8],
+    ) -> McapResult<()> {
+        let chunk_indexes = &self.chunk_indexes[self.cur_chunk_index..];
+        // linear search through our chunk indexes to figure out which one it is. In the common case,
+        // the first chunk index will be right.
+        let Some((i, chunk_index)) = chunk_indexes.iter().enumerate().find(|(_, chunk_index)| {
+            let chunk_start_offset = chunk_index
+                .compressed_data_offset()
+                .expect("chunk data start offset checked in new()");
+            chunk_start_offset == offset
+        }) else {
+            return Err(McapError::UnexpectedChunkDataInserted);
+        };
+        if compressed_data.len() != chunk_index.compressed_size as usize {
+            return Err(McapError::UnexpectedChunkDataInserted);
+        }
+        let uncompressed_size = chunk_index.uncompressed_size as usize;
+        let slot_idx = find_or_make_chunk_slot(&mut self.chunk_slots, offset, uncompressed_size);
+
+        let slot = &mut self.chunk_slots[slot_idx];
+        match chunk_index.compression.as_str() {
+            "" => {
+                slot.buf.resize(uncompressed_size, 0);
+                slot.buf[..].copy_from_slice(compressed_data);
+            }
+            #[cfg(feature = "zstd")]
+            "zstd" => {
+                // decompress zstd into current slot
+                slot.buf.clear();
+                slot.buf.reserve(uncompressed_size);
+                let n =
+                    zstd::zstd_safe::decompress(&mut slot.buf, compressed_data).map_err(|err| {
+                        McapError::DecompressionError(zstd::zstd_safe::get_error_name(err).into())
+                    })?;
+                if n != uncompressed_size {
+                    return Err(McapError::DecompressionError(format!(
+                        "zstd decompression error: expected {uncompressed_size}, got {n}"
+                    )));
+                }
+            }
+            #[cfg(feature = "lz4")]
+            "lz4" => {
+                slot.buf.resize(uncompressed_size, 0);
+                use std::io::Read;
+                let mut decoder = lz4::Decoder::new(std::io::Cursor::new(compressed_data))?;
+                decoder.read_exact(&mut slot.buf[..])?;
```

Comment:
Is there some cost to creating a new decoder every time? Like setting up a thread pool or something?